Adds RAII wrappers similar to vk::raii.

marovira opened the pull request:
This adds a set of wrappers similar to `vk::raii`. 
Fixes #9

Cvelth reviewed (approved):
Looks good! Thanks for taking the time to look into this!

Cvelth commented on include/vkfw/vkfw_raii.hpp:
```diff
@@ -0,0 +1,426 @@
+      VKFW_CONSTEXPR_14 Cursor(Cursor &&) = default;
+      ~Cursor() {
+  #ifdef VKFW_NO_EXCEPTIONS
+        VKFW_NAMESPACE::ignore(m_cursor.destroy());
```
Ah, wait, I just realized that destructor is not supposed to throw, ever, so I guess `ignore` is here to stay, no matter if exceptions are enabled or not.

Which, I guess, means that we don't have to wrap the entire thing into `#ifndef VKFW_NO_EXCEPTIONS`, only the new constructor.

Let's just have a really big disclaimer near the entry for the new macro in `README.md` about the fact that these wrappers end up silently discarding all the destruction errors.

Cvelth commented on include/vkfw/vkfw_raii.hpp:
```diff
@@ -0,0 +1,426 @@
+        : m_window{ VKFW_NAMESPACE::createWindow(width, height, title, hints, monitor, share,
+                                                 reset_hints) } {}
+  #endif
+      VKFW_CONSTEXPR Window(GLFWwindow *window) VKFW_NOEXCEPT : m_window{ window } {}
+      VKFW_CONSTEXPR Window(VKFW_NAMESPACE::Window const &window) VKFW_NOEXCEPT
+        : m_window{ window } {}
```
I just reread the original issue. If we want to guarantee that this `Window` is always in a valid state, we'd have to ensure no `nullptr` (or `nullptr` wrapped into a `vkfw::Window` is passed in.

Which makes me think whether we even want these constructors in the first place. Maybe it's best we just keep the one "main" constructor for these and that is. For all the other use-cases `vkfw::Window` is probably better than `vkfw::raii::Window` anyway.

Cvelth replied:
Let's continue this discussion here so that it's all in the same place:

> I left the constructors of the RAII handles that accept either a `vkfw::` or `GLFW` handle. My reasoning is this: if exceptions are disabled, then there's no way to construct the wrappers and we shouldn't penalise users that wish to use them by disabling them. Yes, this does open up the possibility that a null handle is passed in, but given the code, that would result in a no-op which is fine. The RAII handles from Vulkan-HPP have this same behaviour, so I think it's fine to leave them as is.

I guess if `Vulkan-HPP` behaves the same way we probably want to keep similar behavior at the very least for consistency sake 🤔, even if I highly dislike it.

----------------------------------

> if exceptions are disabled, then there's no way to construct the wrappers and we shouldn't penalise users that wish to use them by disabling them

Yes, but then those same users have no way of knowing something went wrong... I guess we have already accepted the same trade-off for destructors, but IMO, constructor errors are way more important, and common too. But consistency...

Can we at the very least make these "laundering" constructors throw an error or something when a `nullptr` is passed when exceptions are enabled - as they are the main use-case here, exceptionless setups wouldn't have much use for these after all?

----------------------------------

> but given the code, that would result in a no-op which is fine

Wait, what about the cases like this?
```
// (pseudocode)
vkfw::UniqueWindow my_fake_window = nullptr;
vkfw::raii::Window my_laundered_window(std::move(my_fake_window));

my_laundered_window.getPos(); // whoops!
```

marovira commented:
I've revamped the entire RAII system to streamline things and make it easier. I've also resolved all of the previous reviews to minimse noise in the new version.
Here's a quick summary of what I changed:
1. The RAII handles now live in `vkfw.hpp`. I thought about this more and you were right: there's no reason not to have them there.
2. I've added `VKFW_NO_RAII_HANDLE` to disable them, though they will also be disabled if `VKFW_DISABLE_ENHANCED_MODE` is defined.
3. The destructors of the RAII wrappers now behave identically to the destructors of the unique handles. I've also added comments so users can see what's going on.
4. The `README` has been updated to show the new flags and a brief guide of how to use them.
5. I've added a new example that shows how the RAII wrappers can be used.

I think this covers pretty much everything we've discussed, with two notable exceptions:
1. I left the constructors of the RAII handles that accept either a `vkfw::` or `GLFW` handle. My reasoning is this: if exceptions are disabled, then there's no way to construct the wrappers and we shouldn't penalise users that wish to use them by disabling them. Yes, this does open up the possibility that a null handle is passed in, but given the code, that would result in a no-op which is fine. The RAII handles from Vulkan-HPP have this same behaviour, so I think it's fine to leave them as is.
2. I left the list of wrapped functions for `vkfw::raii::Window`. As I explained before, these wrapper functions are intentional because the class itself *is a wrapper*. The duplication of the API is intentional and allows the wrapper to be used as if it were the `vkfw::Window` class. Note that the conversion operators are still there if users want to fallback to the base handles.

Let me know what you think!

Cvelth commented on README.md:
```diff
@@ -143,6 +176,9 @@ The `UniqueHandle` helper class and all the unique handle types and functions re
 > If `VKFW_NO_INCLUDE_VULKAN_HPP` isn't defined, then `vkfw` will use the smart handles from Vulkan-HPP directly. In the event that `VULKAN_HPP_NO_SMART_HANDLES` is defined, then `VKFW_NO_SMART_HANDLE` will also be defined if it hasn't been defined already and a compiler warning will be issued.
 > You can avoid this behavior by defining either `VKFW_NO_SMART_HANDLE` or `VKFW_NO_INCLUDE_VULKAN_HPP` before including `<vkfw/vkfw.hpp>`
 
+### `VKFW_NO_RAII_HANDLE`
+The RAII wrappers and the `raii` namespace (if `VKFW_RAII_NAMESPACE` isn't defined) will be disabled.
```
Please add a note about this having no effect if all the unique handles are disabled.

Cvelth commented on README.md:
```diff
@@ -165,6 +201,10 @@ Removes leading `e` from enum values, for example `vkfw::Key::eA` becomes `vkfw:
 ### `VKFW_NAMESPACE`
 By default, the namespace used with `vkfw.hpp` is `vkfw`. By defining `VKFW_NAMESPACE` before including `vkfw.hpp`, you can adjust this.
 
+### `VKFW_RAII_NAMESPACE`
+By default, the namespace used with the RAII wrappers is `raii`. By defining `VKFW_RAII_NAMESPACE` before including `vkfw.hpp`, you can adjust this.
+Note that the full namespace will be `VKFW_NAMESPACE::VKFW_RAII_NAMESPACE`.
```
Very enterprise!

Cvelth commented on example/empty_raii.cpp:
```diff
@@ -0,0 +1,99 @@
+//
+//========================================================================
+//
+// Conversion to vkfw (and C++): John Cvelth <cvelth.mail@gmail.com>
```
Please add a similar line below about what was modified and by who.

Cvelth commented on include/vkfw/vkfw_raii.hpp:
```diff
@@ -0,0 +1,426 @@
+namespace VKFW_NAMESPACE {
+  namespace VKFW_RAII_NAMESPACE {
+  #ifndef VKFW_NO_EXCEPTIONS
+    class Instance {
```
This thread still applies, what do you think?

Cvelth commented on include/vkfw/vkfw_raii.hpp:
```diff
@@ -0,0 +1,426 @@
+      }
+      VKFW_NODISCARD typename ResultValueType<std::tuple<int, int>>::type getPos() const {
+        return m_window.getPos();
+      }
```
Let's use this thread for this discussion, so it's all in one place.

> I left the list of wrapped functions for `vkfw::raii::Window`. As I explained before, these wrapper functions are intentional because the class itself _is a wrapper_. The duplication of the API is intentional and allows the wrapper to be used as if it were the `vkfw::Window` class. Note that the conversion operators are still there if users want to fallback to the base handles.

Yes, I'm not arguing that we don't want wrapper functions, because we certainly do. What I'm trying to say is that we're very likely to be able to get away with just their names - no prototypes, no bodies, all while achieving the exact same functionality. Please take another look at the godbolt example above.

Cvelth commented on example/empty_raii.cpp:
```diff
@@ -0,0 +1,99 @@
+//========================================================================
+// Empty event test
```
It's a really good idea to add an example! Thanks!

marovira commented:
I'm making this a comment because I want to keep it separate from the rest of the reviews.
I just found something interesting while stepping through this:
```c++
// Assume GLFW has been correctly initialised.
vkfw::Window win{nullptr};
win.getPos();
```
I thought that GLFW would issue an error if the underlying `GLFWwindow*` was null. Turns out it doesn't! Stepping through the code, all they do is have an `assert` that the handle isn't null. If asserts are disabled, then the code succeeds with no error codes being triggered. The implementation in VKFW also doesn't check for a null handle, so ultimately it is possible to have (in all contexts) a null window and perform operations on it that silently do nothing if asserts are disabled.
Is this a good idea? Should this behaviour be changed?

Cvelth commented:
(sorry, I didn't have much time to dedicate to this today, I'll take a look tomorrow or the day after)

Cvelth commented:
@marovira, sorry for the delay! I'm on this now.

> I thought that GLFW would issue an error if the underlying `GLFWwindow*` was null.

Yes, that's exactly why I'm so concerned about error handling in the constructor. Practically, the main way to end up with a `nullptr` is for the, for example, window construction to fail.

> The implementation in VKFW also doesn't check for a null handle, so ultimately it is possible to have (in all contexts) a null window and perform operations on it that silently do nothing if asserts are disabled.

Yes, mostly because it's somewhat hard to end up with null handles. They usually come from "factory"-like functions (at least they used to pre-raii) - and those heavily encourage you to actually take care of the errors they might produce.

In essence, using anything on a null handle is pretty much always a user error. Most likely because they decided to skip error checking.

> Is this a good idea? Should this behaviour be changed?

Do I like it? No. Do I think it's worth it to make best base slower (by adding extra checks) to ensure worst case (improper usage) is easier to debug? Also, no.

I wouldn't be opposed to some _opt-in_ assertions. Think hiding some checks behind a macro like `VKFW_CHECK_FOR_NULL_HANDLES` that's not defined by default, but honestly, I feel like the way to go is to just be extra explicit (in comments, error messages, etc) that it's usually a _really_ bad idea to ignore errors. Especially construction errors.

> Note: if this behaviour isn't ideal, then I propose that the RAII wrappers be merged first and then I'll open a new pull request that fixes that. Mainly so that this can get in before I start changing the code further

Yes, I 100% agree that this is not directly related to the raii wrappers, let's get these in first!

P. S. On the topics of PRs, I just merged another one, could you please rebase?

P. P. S. Also, it looks like there are un-answered threads, could you please take a look when you have some time?

marovira commented:
@Cvelth there's currently 2 unresolved threads: exception management and the explicitness of the RAII wrapper for window. The threads are really messy to read right now, but those are the only ones left.

> P. S. On the topics of PRs, I just merged another one, could you please rebase?

Rebased.

marovira commented:
Just a heads up: starting next week it's looking like I'll have next to no time to work on other projects. If we could get this merged before then that would be great!

Cvelth commented:
> If we could get this merged before then that would be great!

Let's try to do that then!

Cvelth commented on example/empty.cpp:
```diff
@@ -28,6 +28,7 @@
 //========================================================================
 //
 // Conversion to vkfw (and C++): John Cvelth <cvelth.mail@gmail.com>
+// Adapted to use vkfw RAII: Mauricio A. Rovira Galvez
```
Wait, this is not the right file 😅.

Cvelth commented:
> I have left replies to both of threads.

Wait, I don't see the replies 🤔

Cvelth commented on README.md:
````diff
@@ -89,6 +89,39 @@ As all the deleters are known at compiler time and don't need to store anything
+directly like this:
+```c++
+vkfw::raii::Window win{400, 400, "raii_window"};
+// win will destroy the underlying handle when it goes out of scope.
````
Maybe something like
```cpp
// The destructor of `vkfw::raii::Window` will destroy...
```
would be clearer? What do you think?

marovira commented:
> Let's use this thread for this discussion, so it's all in one place.
> 
> > I left the list of wrapped functions for `vkfw::raii::Window`. As I explained before, these wrapper functions are intentional because the class itself _is a wrapper_. The duplication of the API is intentional and allows the wrapper to be used as if it were the `vkfw::Window` class. Note that the conversion operators are still there if users want to fallback to the base handles.
> 
> Yes, I'm not arguing that we don't want wrapper functions, because we certainly do. What I'm trying to say is that we're very likely to be able to get away with just their names - no prototypes, no bodies, all while achieving the exact same functionality. Please take another look at the godbolt example above.

I get what you're saying, but I have two points against it:
1. The conversion/access operators that are there would break or would make no sense if the RAII class derived from the `vkfw` handle. Let me elaborate:
In Vulkan-HPP, the RAII wrappers wrap around the C++ handles (so `vk::raii::Device` wraps `vk::Device`). As a consequence of this, the `vk::raii` wrappers provide a way to access the underlying handle in two ways: `operator*` and `operator vk::<Type>`. Both of these operators effectively return a copy of the underlying handle.
The `vkfw::raii` handles behave the same way (with the exception of `vkfw::raii::Instance`). In the case of `vkfw::raii::Window`, it has both `operator*` and `operator vkfw::Window`. 
Here's the rub: if `vkfw::raii::Window` were to derive from `vkfw::Window`, then `operator vkfw::Window` would break, as there is no way to convert from a class that derives privately into the base class. In the case of `operator*` it would have to cast the instance into the base class and return a copy of the resulting cast. This just feels wrong, since it's bypassing the private inheritance and it is also resulting in slicing the derived class to get the base. For this (simple) case it's maybe fine, but I really dislike this practice as it can lead to surprising bugs.
2. In my opinion/experience, wrappers should never derive from the classes they wrap as it is possible for the base classes to change behaviour and silently break the wrappers that derive from them. I'm aware it is highly unlikely to happen here, but the principle still stands.

In summary, I'm against the inheritance because of what it implies for the wrapper and the breaking of the conversion interface that should be kept to retain compatibility with `vk::raii`. I'm especially against this since all we're really saving is not having to implement the function bodies, since the entire stack of functions (including macro guards for spans and whatnot) would need to be there regardless.

marovira commented:
> @marovira, sorry for the delay! I'm on this now.
> 
> > I thought that GLFW would issue an error if the underlying `GLFWwindow*` was null.
> 
> Yes, that's exactly why I'm so concerned about error handling in the constructor. Practically, the main way to end up with a `nullptr` is for the, for example, window construction to fail.
> 
> > The implementation in VKFW also doesn't check for a null handle, so ultimately it is possible to have (in all contexts) a null window and perform operations on it that silently do nothing if asserts are disabled.
> 
> Yes, mostly because it's somewhat hard to end up with null handles. They usually come from "factory"-like functions (at least they used to pre-raii) - and those heavily encourage you to actually take care of the errors they might produce.
> 
> In essence, using anything on a null handle is pretty much always a user error. Most likely because they decided to skip error checking.
> 
> > Is this a good idea? Should this behaviour be changed?
> 
> Do I like it? No. Do I think it's worth it to make best base slower (by adding extra checks) to ensure worst case (improper usage) is easier to debug? Also, no.
> 
> I wouldn't be opposed to some _opt-in_ assertions. Think hiding some checks behind a macro like `VKFW_CHECK_FOR_NULL_HANDLES` that's not defined by default, but honestly, I feel like the way to go is to just be extra explicit (in comments, error messages, etc) that it's usually a _really_ bad idea to ignore errors. Especially construction errors.
> 
> > Note: if this behaviour isn't ideal, then I propose that the RAII wrappers be merged first and then I'll open a new pull request that fixes that. Mainly so that this can get in before I start changing the code further
> 
> Yes, I 100% agree that this is not directly related to the raii wrappers, let's get these in first!
> 
> P. S. On the topics of PRs, I just merged another one, could you please rebase?
> 
> P. P. S. Also, it looks like there are un-answered threads, could you please take a look when you have some time?

It feels like there might be a misunderstanding on the behaviour of the constructors of the RAII wrappers (could be wrong though). Just in case:
The *base* constructor of the RAII handles (i.e. the only one that is allowed to initialise the underlying `vkfw::` handle) is only enabled when exceptions are enabled. Therefore, if a user tries to create a RAII handle and it somehow fails, an exception is thrown. On the other hand, if exceptions are disabled, then the base constructors are removed, meaning the only way to create a RAII handle is to manually create the `vkfw::` handle separately and then pass it in. In other words, users would have to do something like this:
```c++
auto [code, win] = vkfw::createWindow(...);
if (!code) {
   // Do something
}
vkfw::raii::Window raii{win};
// raii now takes ownership of the window and will destroy it once it goes out of scope
```

The remaining two constructors that either take a `vkfw::` handle or a `GLFW` pointer exist to allow the code I showed above so users can still use the RAII handles if exceptions are disabled (in the same way the unique handles continue to work even when exceptions are disabled). This also conforms with the behaviour of the RAII handles from Vulkan: they all allow taking ownership of a handle that was (by necessity) constructed separately.

> In essence, using anything on a null handle is pretty much always a user error. Most likely because they decided to skip error checking.

This comment still applies. The only way a null handle gets into the RAII wrappers is through user error (either they made a mistake or decided to ignore error codes). The behaviour of the RAII handle is then identical to the behaviour of the base `vkfw::` handle under a null pointer. Vulkan-HPP has something similar: it is perfectly fine to create a RAII handle with a `vk::` object that hasn't been initialised.

I can add an exception to the RAII handles if a null pointer/handle is passed in, but it feels a bit asymmetric since the base handles can happily receive a null handle and not raise errors until they are used.

Cvelth commented:
I typed a bunch stuff here, but I got rid of it.

Thinking about this, the worst thing we could possibly do is introduce inconsistencies with `VulkanHPP`. So, yeah, you're right, let's keep it as close to their implementation as possible. Which, from the looks of it, it already is. Good job!

The only comment I'll keep is:

> I can add an exception to the RAII handles if a null pointer/handle is passed in, but it feels a bit asymmetric since the base handles can happily receive a null handle and not raise errors until they are used.

No, all the handles could use this. Copying a `null` handle is a bug in 99% of cases. But again, yeah, as you have pointed out, it's still a user-error, so not the biggest of deals.

----------------------------------------

So, yeah, on the fresh look, everything looks good, thanks for doing such a good job (and convincing me that what I'm proposing sucks! :D).

The only thing that's missing, I'd say, is to add new options to the CMake, in case someone would prefer to specify them from there, I mean [this](https://github.com/Cvelth/vkfw/blob/main/CMakeLists.txt#L24) and [this](https://github.com/Cvelth/vkfw/blob/main/CMakeLists.txt#L77) lists.

marovira commented:
> The only thing that's missing, I'd say, is to add new options to the CMake, in case someone would prefer to specify them from there, I mean [this](https://github.com/Cvelth/vkfw/blob/main/CMakeLists.txt#L24) and [this](https://github.com/Cvelth/vkfw/blob/main/CMakeLists.txt#L77) lists.

Now I feel sheepish. You'd think I remember seeing as I added the CMake code... but nope!

> So, yeah, on the fresh look, everything looks good, thanks for doing such a good job (and convincing me that what I'm proposing sucks! :D).

No worries! It's fun having debates like these.
